File: ruby/list2-10.rb
#1,000以下の素数を列挙する list2-9.rbの改正版

i = 2
counter = 0

2.upto(1000) do |n|

  while i < n             # 自分以外の数字で除算する
    counter += 1          # 除算回数カウント
    break if n%i == 0     # 割り切れたら除算終了 -> 素数ではない
    i += 1                # 割る数加算
  end

  puts n if i == n        # while文をfalseで抜けると素数なので出力

  i = 2   #iを2に初期化 nを2から除算開始するため

end

puts "除算の回数：#{counter}"

File: c/list2-10.c
//1,000までの素数を列挙する list2-9の最適化
//素数：1より大きい自然数で、1とその数自身以外のどのような自然数でも割り切れない数
//計算数14622回

#include<stdio.h>

int main(void)
{
  int i, n;
  int prime[167];                 //list2-9.cで得られた素数の数
  int num = 0;
  unsigned long counter = 0;

  prime[num++] = 2;               //2は素数
  prime[num++] = 3;               //3は素数

  for (n = 5; n <= 1000; n += 2)  //対象の数は奇数のみ、2以上の偶数は素数でないのは明らか
  {
    for (i = 1; i < num; i++)
    {
      counter++;
      if (n%prime[i] == 0)        //対象の数以下の素数で割り切れたら素数じゃない
      {
        break;
      }
    }
    if (num == i)                 //for文最後までまわれば(割り切れてない)、対象の数は素数
    {
      prime[num++] = n;           //配列に素数を追加
    }
  }

  for (int i = 0; i < num; i++)
  {
    printf("%d\n", prime[i]);
  }

  printf("除算の回数：%lu\n", counter);   //-> 14622回

  return 0;
}